server/controller_client.go

```diff
@@ -0,0 +1,231 @@
+		go c.startStream(unit)
+	}
+}
+
+// handle the UnitChangedModified event from the V2 API
+func (c *clientHandler) handleUnitUpdated(unit *client.Unit) {
+	c.log.Debugf("Got Unit Modified: %s", unit.ID())
+	unitType := unit.Type()
+
+	if unitType == client.UnitTypeOutput {
+		state, _ := unit.Expected()
+		if state == client.UnitStateStopped {
+			c.shutdown(unit)
+		}
+	} else {
+		c.updateStream(unit)
+	}
+}
+
+// a blocking call that will wait for the shipper components to gracefully shutdown, then send the unit update
+func (c *clientHandler) shutdown(shipperUnit *client.Unit) {
+	if c.shipperIsStopping {
+		return
+	}
+	_ = shipperUnit.UpdateState(client.UnitStateStopping, "shutting down shipper output", nil)
+	c.stopShipper()
+	// The server has successfully shut down
+	// In theory we want this to block, as we should wait for queues & outputs to empty.
+	c.shutdownComplete.Wait()
+	c.log.Debugf("Shutdown complete, sending STOPPPED")
+	_ = shipperUnit.UpdateState(client.UnitStateStopped, "Stopped shipper output", nil)
+}
+
+// runController is the main runloop for the shipper itself, and managed communication with the agent.
+func runController(ctx context.Context, agentClient client.V2) error {
+	log := logp.L()
+
+	err := agentClient.Start(ctx)
+	if err != nil {
+		return fmt.Errorf("error starting connection to client")
+	}
+
+	log.Debugf("Starting error reporter")
+	go reportErrors(ctx, agentClient)
+
+	log.Debugf("Started client, waiting")
+
+	handler := newClientHandler()
+
+	// receive the units
+	for {
+		select {
+		case <-ctx.Done():
+			handler.log.Debugf("Got context done")
+			shipperUnit := handler.getUnit(handler.shipperOutputID)
+			handler.shutdown(shipperUnit)
+			// If we get context done, just hard-stop
+			return nil
+
+		case change := <-agentClient.UnitChanges():
+
+			switch change.Type {
+			case client.UnitChangedAdded: // The agent is starting the shipper, or we added a new processor
+				go handler.handleUnitAdded(change.Unit)
+			case client.UnitChangedModified: // config for a unit has changed
+				go handler.handleUnitUpdated(change.Unit)
+			case client.UnitChangedRemoved: // a unit has been stopoped and can now be removed.
+				handler.deleteUnit(change.Unit)
+				// for now, consider a remove of the shipper unit to be our sign to shut down.
+				// Take care, as we won't get this until we've sent a STOPPED to the agent
+				// TODO: we should have a timeout so we can shutdown without getting a REMOVED event
+				if change.Unit.Type() == client.UnitTypeOutput {
+					handler.log.Debugf("shipper unit removed, ending.")
+					return nil
+				}
+
+			}
+		}
+	}
+}
+
+// I am not net sure how this should work or what it should do, but we need to read from that error channel
+func reportErrors(ctx context.Context, agentClient client.V2) {
+	log := logp.L()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case err := <-agentClient.Errors():
```

Comment:
Same here, there must be a guarantee that the channel will always unblock, it this true?

Reply:
I assume this is also a runtime loop until the process exists, so it's fine to read from this channel.